command/sub8_missions/missions/collect_classify.py

```diff
@@ -0,0 +1,395 @@
+        then submerge again.
+        '''
+        yield self.sub.move.depth(0).go()
+        yield self.sub.nh.sleep(2.0).go()
+        depth = self.height_to_depth(CollectClassifyMission.OVAL_SEARCH_HEIGHT)
+        yield self.sub.move.depth(depth).go()
+
+    @util.cancellableInlineCallbacks
+    def circle(self, origin, radius, num_moves=10, proportion=1.0):
+        '''
+        Execute discrete moves to circle around a point while looking at the point.
+        '''
+        depth = -origin[2]
+        start_move = self.sub.move.look_at(origin).set_position(origin).backward(radius).depth(depth)
+        start_position = start_move._pose.position
+        yield start_move.go()
+        offset = start_position - origin
+        start_theta = np.arctan2(offset[1], offset[0])
+        for theta in np.linspace(start_theta, proportion * 2.0 * np.pi + start_theta, num=num_moves):
+            d = np.array([radius * np.cos(theta), radius * np.sin(theta), 0])
+            p = origin + d
+            yield self.sub.move.set_position(p).look_at_without_pitching(self.pinger_position).go()
+
+    @util.cancellableInlineCallbacks
+    def find_pipe(self):
+        '''
+        Circle around pinger, looking for pipes in the tower. If one is found,
+        return the Pipe object. Otherwise, return None.
+        '''
+        yield self.sub.vision_proxies.tower_pipes.start()
+        circle_done = False
+
+        @util.cancellableInlineCallbacks
+        def circle_pinger():
+            origin = self.pinger_position
+            origin[2] = -self.height_to_depth(self.PIPE_SEARCH_HEIGHT)
+            yield self.circle(origin, CollectClassifyMission.PIPE_SEARCH_RADIUS)
+            self.circle_done = True
+        search = circle_pinger()
+        while not circle_done:
+            res = self.get_pipe()
+            if res is not None:
+                # TODO: check frame, transform to map if needed
+                circle_done = True
+                search.cancel()
+                yield self.sub.vision_proxies.tower_pipes.stop()
+                defer.returnValue(res)
+        yield self.sub.vision_proxies.tower_pipes.stop()
+        defer.returnValue(None)
+
+    @util.cancellableInlineCallbacks
+    def pickup_pipe(self, pose):
+        '''
+        Execute moves to grab a pipe (given its pose)
+        '''
+        # Move behind
+        move = self.sub.move.set_position(pose[0]).set_orientation(pose[1]).yaw_left(
+            np.pi / 2.0).zero_roll_and_pitch()
+        # Adjust for gripper tf
+        move = move.relative(CollectClassifyMission.BASE_LINK_TO_ARM)
+        # Rotate 180 if we're not facing pinger
+        m = trns.quaternion_matrix(self.sub.move._pose.orientation).dot(np.array([0.1, 0, 0]))
+        if np.linalg.norm(m - self.pinger_position) > np.linalg.norm(move._pose.position - self.pinger_position):
+            move = move.yaw_left(np.pi)
+
+        move = move.backward(1.5)
```

Comment:
I think a lot of these moves should be done with a non-default speed. Things get scary when the sub is moving as fast as it can. What do you think?